From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kanaka Raju Nayana <kanaka.raju.nayana@intel.com>
Date: Wed, 12 Aug 2026 13:46:33 +0530
Subject: [PATCH] mei: use non-const class/bus descriptors on older kernels

driver core constified class_register() and bus_register() in v6.4,
allowing struct class and struct bus_type to reside in read-only memory.
On kernels older than v6.4, writing into these const objects during
registration causes a page fault in __class_register.

Control mei_class and mei_cl_bus_type declarations so non-const is used
when the kernel does not support const registration arguments.

Signed-off-by: Kanaka Raju Nayana <kanaka.raju.nayana@intel.com>
---
 drivers/misc/mei/bus.c  | 11 +++++++++++
 drivers/misc/mei/main.c |  6 ++++++
 2 files changed, 17 insertions(+)

diff --git a/drivers/misc/mei/bus.c b/drivers/misc/mei/bus.c
index fcde082..837fd34 100644
--- a/drivers/misc/mei/bus.c
+++ b/drivers/misc/mei/bus.c
@@ -1300,6 +1300,16 @@ static int mei_cl_device_uevent(const struct device *dev, struct kobj_uevent_env
 	return 0;
 }
 
+#ifdef BPM_MEI_CLASS_BUS_REGISTER_CONST_ARGS_NOT_PRESENT
+static struct bus_type mei_cl_bus_type = {
+        .name           = "mei",
+        .dev_groups     = mei_cldev_groups,
+        .match          = mei_cl_device_match,
+        .probe          = mei_cl_device_probe,
+        .remove         = mei_cl_device_remove,
+        .uevent         = mei_cl_device_uevent,
+};
+#else
 static const struct bus_type mei_cl_bus_type = {
 	.name		= "mei",
 	.dev_groups	= mei_cldev_groups,
@@ -1308,6 +1318,7 @@ static const struct bus_type mei_cl_bus_type = {
 	.remove		= mei_cl_device_remove,
 	.uevent		= mei_cl_device_uevent,
 };
+#endif
 
 static struct mei_device *mei_dev_bus_get(struct mei_device *bus)
 {
diff --git a/drivers/misc/mei/main.c b/drivers/misc/mei/main.c
index 54f70f5..4f7f402 100644
--- a/drivers/misc/mei/main.c
+++ b/drivers/misc/mei/main.c
@@ -29,9 +29,15 @@
 #include "mei_dev.h"
 #include "client.h"
 
+#ifdef BPM_MEI_CLASS_BUS_REGISTER_CONST_ARGS_NOT_PRESENT
+static struct class mei_class = {
+        .name = "mei",
+};
+#else
 static const struct class mei_class = {
 	.name = "mei",
 };
+#endif
 
 static dev_t mei_devt;
 #define MEI_MAX_DEVS  MINORMASK
-- 
2.43.0
